### PR TITLE
Devoncarew test 0 12 0

### LINKED ISSUE #58: Integrate with the `test` test runner
Integrate the Dart Test Runner (https://github.com/google/test_runner.dart) in some way.

### LINKED ISSUE #217: upgrade to unit test 0.12.0

### PR DESCRIPTION
**Pull request by devoncarew**
fix #217 
fix #58 

- add a wrapper for the `test` package, called `Test`
- deprecate the `Tests` class
- import both `test` and `unittest` so we can deprecate functionality before removing it
- switch our unit tests over to using `test` and dogfood `Test` in our build script

@seaneagan

### CODE REVIEW

**Comment by zoechi** [comment]
I was thinking about migrating the tests to the new `test` package but you would loose coverage. Coverage doesn't work with `test` yet (AFAIK).

File: lib/grinder_tools.dart [hunk 1]
```diff
@@ -175,6 +176,58 @@ class Tests {
...
 
+/// A wrapper around the `test` package. This class is used to run your unit
+/// tests.
+class Test {
+  final PubApp _test = new PubApp.local('test');
+
+  Test();
+
+  /// Run the tests in the current package. See
+  /// https://pub.dartlang.org/packages/test.
```
Comment (seaneagan) [inline lib/grinder_tools.dart]: will dartdoc make this link clickable?  Maybe use a markdown link:

the [test package](https://pub.dartlang.org/packages/test).

File: lib/grinder_tools.dart [hunk 2]
```diff
@@ -175,6 +176,58 @@ class Tests {
...
+  /// [name] is substring of the name of the test to run. Regular expression
+  /// syntax is supported. [plainName] is a plain-text substring of the name of
+  /// the test to run. [platform] is the platform(s) on which to run the tests.
+  /// Available values are `vm` (default), `dartium`, `content-shell`, `chrome`,
+  /// `phantomjs`, `firefox`, `safari`. [concurrency] controls the number of
+  /// concurrent test suites run (defaults to 4). [pubServe] is the port of a
+  /// pub serve instance serving `test/`.
+  void test({String name, String plainName, List<String> platforms,
+      int concurrency, int pubServe, RunOptions runOptions}) {
+    List<String> args = ['--reporter=expanded'];
```
Comment (seaneagan) [inline lib/grinder_tools.dart]: Move this and the next five lines to a utility function that is used here and in `testAsync`?

File: lib/grinder_tools.dart [hunk 3]
```diff
@@ -175,6 +176,58 @@ class Tests {
...
+  /// output uses ANSI colors.
+  Future testAsync({String name, String plainName, List<String> platforms,
+      int concurrency, int pubServe, bool color, RunOptions runOptions}) {
+    List<String> args = ['--reporter=expanded'];
+    if (name != null) args.add('--name=${name}');
+    if (plainName != null) args.add('--plain-name=${plainName}');
+    if (platforms != null) args.add('--platform=${platforms.join(',')}');
+    if (concurrency != null) args.add('--concurrency=${concurrency}');
+    if (pubServe != null) args.add('--pub-serve=${pubServe}');
+    if (color != null) args.add(color ? '--color' : '--no-color');
```
Comment (seaneagan) [inline lib/grinder_tools.dart]: Why is `color` only supported for `testAsync` and not `test`?

File: lib/grinder_tools.dart [hunk 4]
```diff
@@ -175,6 +176,58 @@ class Tests {
...
+    _test.run(args, script: 'test', runOptions: runOptions);
+  }
+
+  /// Run the tests in the current package. See
+  /// https://pub.dartlang.org/packages/test.
+  ///
+  /// [name] is substring of the name of the test to run. Regular expression
+  /// syntax is supported. [plainName] is a plain-text substring of the name of
+  /// the test to run. [platform] is the platform(s) on which to run the tests.
+  /// Available values are `vm` (default), `dartium`, `content-shell`, `chrome`,
```
Comment (seaneagan) [inline lib/grinder_tools.dart]: Can we just link to the test packages documentation on this instead of inlining it here:

https://github.com/dart-lang/test#platform-selector-syntax

It would also be nice to link to test package for the other docs as well, but those don't seem to have documentation in the README.

File: lib/grinder_tools.dart [hunk 5]
```diff
@@ -175,6 +176,58 @@ class Tests {
...
+  /// `phantomjs`, `firefox`, `safari`. [concurrency] controls the number of
+  /// concurrent test suites run (defaults to 4). [pubServe] is the port of a
```
Comment (seaneagan) [inline lib/grinder_tools.dart]: Can we just link here:

https://github.com/dart-lang/test#testing-with-barback

File: lib/grinder_tools.dart [hunk 6]
```diff
@@ -175,6 +176,58 @@ class Tests {
...
+  /// pub serve instance serving `test/`.
+  void test({String name, String plainName, List<String> platforms,
```
Comment (seaneagan) [inline lib/grinder_tools.dart]: I think we should support `platforms` being a `String` either instead of or in addition to `List<String>` to get the full power of the [platform selector syntax](https://github.com/dart-lang/test#platform-selector-syntax).  I think once https://github.com/dart-lang/test/issues/23 is fixed, it will probably expose an enum instead, so that'd be one reason to maybe leave out the `List<String>` option for now until that shakes out.

Comment (seaneagan) [inline lib/grinder_tools.dart]: also maybe `platformSelector` may be more self-documenting in at least in the case of `String`.

File: lib/grinder_tools.dart [hunk 7]
```diff
@@ -175,6 +176,58 @@ class Tests {
   }
 }
 
+/// A wrapper around the `test` package. This class is used to run your unit
+/// tests.
+class Test {
```
Comment (seaneagan) [inline lib/grinder_tools.dart]: I like that `Test` matches the name of the package (`test`), but without reading the documentation for it, I'd assume `Test` represents an individual unit test, or run thereof, not a test running API.  Also once https://github.com/dart-lang/test/issues/23 is fixed, that will expose just such a `Test` class, and that may become the preferred way to run tests from grinder since it avoid shelling out, but we would likely have a name conflict.  Some other suggestions would be `Tester`, `TestRunner`, or `TestApp`.  @zoechi thoughts?

File: lib/grinder_tools.dart [hunk 8]
```diff
@@ -61,6 +61,7 @@ void defaultClean([GrinderContext context]) => delete(buildDir);
 /**
  * A utility class to run tests for your project.
  */
+@Deprecated('see `Test`')
```
Comment (seaneagan) [inline lib/grinder_tools.dart]: I thnk we should start using e.g. `'see [Test]'` instead pending the outcome of:

https://github.com/dart-lang/dartdoc/issues/500

but that can be a separate PR to update all of our deprecation messages.

File: lib/grinder_tools.dart (hunk 3 above)
Comment (seaneagan) [inline lib/grinder_tools.dart]: Also, we could look at adding a --color option to grinder itself, and default this one to that, or maybe even remove this one?

Comment (seaneagan) [inline lib/grinder_tools.dart]: See #243.

File: lib/grinder_tools.dart (hunk 7 above)
Comment (zoechi) [inline lib/grinder_tools.dart]: I think `TestRunner` is fine.

File: lib/grinder_tools.dart (hunk 1 above)
Comment (devoncarew) [inline lib/grinder_tools.dart]: done

File: lib/grinder_tools.dart (hunk 2 above)
Comment (devoncarew) [inline lib/grinder_tools.dart]: sgtm

File: lib/grinder_tools.dart (hunk 3 above)
Comment (devoncarew) [inline lib/grinder_tools.dart]: Ah, an oversight. I like your idea of keying off a global color property. I'll remove this param.

File: lib/grinder_tools.dart (hunk 4 above)
Comment (devoncarew) [inline lib/grinder_tools.dart]: I added the link, but want to keep the inlined info, which I think will help users use the class.

File: lib/grinder_tools.dart (hunk 5 above)
Comment (devoncarew) [inline lib/grinder_tools.dart]: I think this will be fairly infrequently used, I just want to expose it for the sake of completeness.

File: lib/grinder_tools.dart (hunk 6 above)
Comment (devoncarew) [inline lib/grinder_tools.dart]: Cool, changed to `dynamic platformSelector`; it can take a list or a string.

File: lib/grinder_tools.dart (hunk 7 above)
Comment (devoncarew) [inline lib/grinder_tools.dart]: kk, `TestRunner` sounds good; I'll document that it maps to the `test` package.

Comment (zoechi) [inline lib/grinder_tools.dart]: but `Test` is fine for me too. No strong opinion here.

File: lib/grinder_tools.dart (hunk 8 above)
Comment (devoncarew) [inline lib/grinder_tools.dart]: Makes sense; changed to `@Deprecated('see [TestRunner]')`.

**Comment by devoncarew** [comment]
Updated - thanks for the comments!

**Comment by seaneagan** [comment]
lgtm!

Can we also now deprecate/remove the browser driving code in grinder?

**Comment by devoncarew** [comment]
> Can we also now deprecate/remove the browser driving code in grinder?

sounds good; let's deprecate until 0.8.0
